### PR TITLE
feat: add automation for ubuntu/* branches asserting quilt patches apply

### PR DESCRIPTION
**Pull request by blackboxsw**
## Proposed Commit Message
`feat: add automation for ubuntu/* branches asserting quilt patches apply`

## Additional Context
[Sample run success](https://github.com/blackboxsw/cloud-init/actions/runs/10398640974/job/28796270466?pr=24)

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: .github/workflows/packaging-tests.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+name: Integration Tests
```
Comment (holmanb) [inline .github/workflows/packaging-tests.yml:1]: It looks like this name didn't get updated from the one this was copied from?

File: .github/workflows/packaging-tests.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+defaults:
+  run:
+    shell: sh -ex {0}
+
+env:
+  RELEASE: focal
+
+jobs:
+  daily-ppa-recipe-check:
```
Comment (holmanb) [inline .github/workflows/packaging-tests.yml:20]: While this isn't necessarily incorrect, the daily PPA isn't the only thing to benefit from this CI job. How about something more generic, like `check-patches`? Also, a `name` key would produce a more user-friendly output, such as `name: Check Patches`

File: .github/workflows/packaging-tests.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    runs-on: ubuntu-22.04
+    steps:
+      - name: "Checkout"
```
Comment (holmanb) [inline .github/workflows/packaging-tests.yml:23]: nit: quotes not necessary

File: .github/workflows/packaging-tests.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        uses: actions/checkout@v3
+        with:
+          # Fetch all branches for merging
+          fetch-depth: 0
+      - name: "Prepare dependencies"
```
Comment (holmanb) [inline .github/workflows/packaging-tests.yml:28]: nit: quotes not necessary

File: .github/workflows/packaging-tests.yml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        run: |
+          sudo DEBIAN_FRONTEND=noninteractive apt-get update
+          sudo DEBIAN_FRONTEND=noninteractive apt-get -y install tox quilt
+      - name: "Setup quilt environment"
```
Comment (holmanb) [inline .github/workflows/packaging-tests.yml:32]: nit: quotes not necessary

File: .github/workflows/packaging-tests.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        run: |
+          echo 'QUILT_PATCHES=debian/patches' >> ~/.quiltrc
+          echo 'QUILT_SERIES=debian/patches/series' >> ~/.quiltrc
+
+      - name: 'Daily recipe: quilt patches apply successfully and tests run'
```
Comment (holmanb) [inline .github/workflows/packaging-tests.yml:37]: nit: quotes not necessary

**Review by holmanb — COMMENTED** [review]
Thanks for putting this together @blackboxsw! I left some comments below, but this looks like a nice addition!

**Comment by holmanb** [comment]
Another job that might be useful is a test that runs when a PR merges which would do the following:

for each series (focal, jammy, noble, devel)

1. check out ubuntu/$SERIES
2. merge main
3. apply patches

This give us the exact commit which introduces a quilt patch conflict which could be useful. Thoughts?

**Comment by blackboxsw** [comment]
> Another job that might be useful is a test that runs when a PR merges which would do the following:
> 
> for each series (focal, jammy, noble, devel)
> 
>     1. check out ubuntu/$SERIES
> 
>     2. merge main
> 
>     3. apply patches
> 
> 
> This give us the exact commit which introduces a quilt patch conflict which could be useful. Thoughts?

@holmanb thanks for the swift review here. This next step you mention is a potential I would like to discuss too. Even a step before trying to create a workflow for each PR against main, it may actually to useful to adapt this workflow to perform the merge in the packaging branch here too. I've pushed a supplemental commit we can discuss. Here's the [successful test run](https://github.com/blackboxsw/cloud-init/actions/runs/10404323488/job/28812549881)

**Review by a-dubs — APPROVED** [review]
Looks good. This seems like a great starting point and we can build out some more advanced automations in the near future!

**Review by holmanb — COMMENTED** [review]
@blackboxsw thanks for the addition. The new addition looks good to me.

> Here's the [successful test run](https://github.com/blackboxsw/cloud-init/actions/runs/10404323488/job/28812549881)

But CI is still failing on this PR?

**Comment by blackboxsw** [comment]
> @blackboxsw thanks for the addition. The new addition looks good to me.
> 
> > Here's the [successful test run](https://github.com/blackboxsw/cloud-init/actions/runs/10404323488/job/28812549881)
> 
> But CI is still failing on this PR?

@holmanb not sure what failed there. Re-kicking the failed job in CI resolved the issue.

**Review by TheRealFalcon — APPROVED** [review]
Great addition!

**Pull request merged**